src/util/netif.cpp

```diff
@@ -0,0 +1,300 @@
+// Copyright (c) 2024 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php.
+
+#include <config/bitcoin-config.h> // IWYU pragma: keep
+
+#include <util/netif.h>
+
+#include <logging.h>
+#include <netbase.h>
+#include <util/check.h>
+#include <util/sock.h>
+#include <util/syserror.h>
+
+// Linux and FreeBSD 13.2+
+#if defined(__linux__) || (defined(__FreeBSD__) && __FreeBSD_version >= 1302000)
+
+#if defined(__linux__)
+#include <linux/rtnetlink.h>
+#elif defined(__FreeBSD__)
+#include <netlink/netlink.h>
+#include <netlink/netlink_route.h>
+#endif
+
+static std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
+{
+    // Create a netlink socket.
+    const int s{socket(AF_NETLINK, SOCK_DGRAM, NETLINK_ROUTE)};
+    if (s < 0) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "socket(AF_NETLINK): %s\n", NetworkErrorString(errno));
+        return std::nullopt;
+    }
+    Sock sock{static_cast<SOCKET>(s)};
```

Comment:
@vasild I haven't had a chance to thoroughly review the second commit. Making a separate PR could help.

Reply:
SGTM.

> That address_family is the first argument to the socket(2) syscall. What's needed is to extend CreateSock*() with the other two arguments - type and protocol. I can do that.

Yes, for specifying UDP (and NETLINK, but i do not intend to make a test framework for that) it would need all three arguments.

Reply:
Done in the topmost commit in https://github.com/vasild/bitcoin/commits/extend_CreateSock/.
Separate PR or include in this PR?

Reply:
This PR is already quite large, so let's make a fresh one that this can rebase on if needed.

Reply:
Agree adding fuzzing/testing makes sense as a follow-up PR. It's already big enough, and also i don't want to interfere with review by doing more active development here than address review comments.

Reply:
Opened https://github.com/bitcoin/bitcoin/pull/30202 "netbase: extend CreateSock() to support creating arbitrary sockets"

Reply:
> Making a separate PR could help

Done, the first two commits from https://github.com/bitcoin/bitcoin/pull/26812 extracted into a separate PR: https://github.com/bitcoin/bitcoin/pull/30205

Reply:
I got a bit confused about these two new PR's, but I guess they are orthogonal.